File: pkmnapi-api/src/requests/trades.rs
use serde::Deserialize;

use crate::requests::base::BaseRequest;

pub type TradeRequest = BaseRequest<TradeRequestType, TradeRequestAttributes>;

impl TradeRequest {
    pub fn get_give_pokedex_id(&self) -> u8 {
        self.data.attributes.give.id
    }

    pub fn get_get_pokedex_id(&self) -> u8 {
        self.data.attributes.get.id
    }
    pub fn get_nickname(&self) -> &String {
        &self.data.attributes.nickname
    }
}

#[derive(Debug, Deserialize)]
#[allow(non_camel_case_types)]
pub enum TradeRequestType {
    trades,
}

#[derive(Debug, Deserialize)]
pub struct TradeRequestAttributes {
    pub give: TradeRequestAttributesPokemon,
    pub get: TradeRequestAttributesPokemon,
    pub nickname: String,
}

#[derive(Debug, Deserialize)]
pub struct TradeRequestAttributesPokemon {
    #[serde(deserialize_with = "crate::utils::from_numeric_str")]
    pub id: u8,
}

File: clients/js/src/requests/trades.ts
import { BaseRequest } from './base';

export type TradeRequest = BaseRequest<TradeRequestType, TradeRequestAttributes>;

enum TradeRequestType {
    trades = 'trades',
}

export interface TradeRequestAttributes {
    give: TradeRequestAttributesPokemon;
    get: TradeRequestAttributesPokemon;
    nickname: string;
}

export interface TradeRequestAttributesPokemon {
    id: string;
}

export const TradeRequestDefault = (attributes: TradeRequestAttributes): TradeRequest => {
    return {
        data: {
            type: TradeRequestType.trades,
            attributes,
        },
    };
};
